[WebDriver][win] make the `New Window` command duplicates an application window, instead of creating new one.

haruhisa-shin opened the pull request:
#### 21275f0da88bfb303cdff0e9681bcae7b8947f84
<pre>
[WebDriver][win] make the `New Window` command duplicates an application window, instead of creating new one.
<a href="https://bugs.webkit.org/show_bug.cgi?id=289887">https://bugs.webkit.org/show_bug.cgi?id=289887</a>

Reviewed by Fujii Hironori.

Before this change, the WebDriver creates new WebPageProxy and WebView for the `New Window` command.
This method is simple but not appropriate for testing the application, as the newly created window
differs from the application&apos;s.
The purpose of this change is to solve the above problem.

The main change in this commit is that the AutomationSessionClient will no longer create a WebView
when it receives `New Window` command, but will instead call the `WKPageUIClient::createNewPage`.
The creation of the window is delegated to the application, which will allow for more correct testing.

Combined changes:
* Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
(WebKit::AutomationClient::closeAutomationSession):
* Source/WebKit/UIProcess/win/AutomationSessionClientWin.cpp:
(WebKit::AutomationSessionClient::requestNewPageWithOptions):
(WebKit::getRootWindowHandle):
(WebKit::AutomationSessionClient::requestMaximizeWindowOfPage):
(WebKit::AutomationSessionClient::requestHideWindowOfPage):
(WebKit::AutomationSessionClient::requestRestoreWindowOfPage):
(WebKit::AutomationSessionClient::close): Deleted.
(WebKit::AutomationSessionClient::didReceiveAuthenticationChallenge): Deleted.
(WebKit::AutomationSessionClient::retainWebView): Deleted.
(WebKit::AutomationSessionClient::releaseWebView): Deleted.
* Source/WebKit/UIProcess/win/AutomationSessionClientWin.h:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
(WebKitBrowserWindow::WebKitBrowserWindow):
(WebKitBrowserWindow::didReceiveAuthenticationChallenge):
(WebKitBrowserWindow::close):
* Tools/MiniBrowser/win/WebKitBrowserWindow.h:

Canonical link: <a href="https://commits.webkit.org/293505@main">https://commits.webkit.org/293505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81e0f0fb1d5473dbb8067ba195d75d7c7bd358f0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103771 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49235 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100691 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18573 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26731 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75100 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32245 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101651 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14095 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89088 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55457 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13876 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7053 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48617 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83830 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7131 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106143 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18760 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84072 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26114 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85289 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83557 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28198 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode; Passed layout tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5879 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19441 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16119 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25695 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30877 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25513 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28833 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27088 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebKit/UIProcess/win/WebPageProxyWin.cpp:
```diff
@@ -85,4 +90,26 @@ void WebPageProxy::dispatchPendingCharEvents(const NativeWebKeyboardEvent& keydo
+    Ref configuration = this->configuration().copy();
+
+    WebCore::WindowFeatures windowFeatures;
+    // FIXME: Attributes of this window should be set to windowFeatures.
```
Please add a bugzilla link

haruhisa-shin replied:
Add the link to https://webkit.org/b/290979.

burg commented on Source/WebKit/UIProcess/win/AutomationClientWin.h:
```diff
@@ -48,19 +46,10 @@ class AutomationSessionClient final : public API::AutomationSessionClient {
-    HashSet<Ref<WebView>> m_webViews;
 };
 
 class AutomationClient final : public API::AutomationClient, Inspector::RemoteInspector::Client {
```
It would be appreciated if, at some point, you could move `AutomationClient` and `AutomationSessionClient` into separate headers and cpp files. They are separate concepts and AutomationSessionClient is responsible for delegating session actions.

haruhisa-shin replied:
I created another pull request only for file separation and fix it first.
https://github.com/WebKit/WebKit/pull/43529

burg commented on Source/WebKit/UIProcess/win/AutomationClientWin.h:
```diff
@@ -48,19 +46,10 @@ class AutomationSessionClient final : public API::AutomationSessionClient {
     void requestNewPageWithOptions(WebKit::WebAutomationSession&, API::AutomationSessionBrowsingContextOptions, CompletionHandler<void(WebKit::WebPageProxy*)>&&) override;
```
Please add this comment above the overrides:
```
     // From API::AutomationSessionClient
```

haruhisa-shin replied:
Added.

burg commented on Source/WebKit/UIProcess/win/WebPageProxyWin.cpp:
```diff
@@ -85,4 +90,26 @@ void WebPageProxy::dispatchPendingCharEvents(const NativeWebKeyboardEvent& keydo
+    auto userInitiatedActivity = API::UserInitiatedAction::create();
+    Ref navigationAction = API::NavigationAction::create(WTFMove(navigationActionData), nullptr, nullptr, String(), WTFMove(request), URL(), false , WTFMove(userInitiatedActivity));
+
+    m_uiClient->createNewPage(*this, WTFMove(configuration), WTFMove(navigationAction), [completionHandler = WTFMove(completionHandler)](auto&& page) mutable {
```
So on first glance, this change seems OK. But comparing the implementation to the other ports, the design is different here, so I want to make sure that any differences are intentional.

As written, the session calls out to the existing PageUIClient to create a new page. As far as I can tell, there's no way for the WebKit client to distinguish this "create new page" request as something requested by a WebDriver command, or if it came from JS, or from WebKit API. I'm concerned that this doesn't account for the rest of what the WebDriver spec is going to need to ask the host app to do.

In the Cocoa ports we have `_WKAutomationSessionDelegate.h` which is an ObjC wrapper around `APIAutomationSessionClient`. This allows for different behavior on iOS and macOS Safari, say, when switching tabs.

In the GTK/WPE ports it seems most of the APIAutomationSessionClient methods are stubbed out with a FIXME to expose an API if the client needs to customize it. If you don't actually need hide/restore/maximize, this seems like the way to go.

haruhisa-shin replied:
At this point, there is no need for the windows ports to customize its behavior on the application side, as is the case with Delegate in the Cocoa ports, so I left this as is.
The hide/restore/maximize feature was added because it was feasible to implement.

burg commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -50,58 +50,25 @@ AutomationSessionClient::AutomationSessionClient(const String& sessionIdentifier
-    retainWebView(WTFMove(newWindow));
-
-    completionHandler(newPage);
+    RefPtr<WebProcessPool> processPool = session.protectedProcessPool();
```
Yeah, this type of thing is not going to work in general, there's no requirement that a process pool contains any pages when a WebDriver session begins.

So if it is the way to handle "New Window" on PlayStation, it needs to be delegated to the client side. But it won't work in general.

haruhisa-shin replied:
In windows ports, WebDriver uses the same protocol as WebInspector, and it should be expected that WebProcess and WebPage are always present in the socket connection.
Therefore, I have not made any changes here.

burg commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -729,6 +729,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     bool isControlledByAutomation() const { return m_controlledByAutomation; }
     void setControlledByAutomation(bool);
+#if PLATFORM(WIN)
+    void requestNewWindow(CompletionHandler<void(RefPtr<WebPageProxy>&&)>&&);
```
I don't think this API is appropriate for WebPageProxy. WebPageProxy doesn't offer other API to create windows vs tabs. The WebKit client needs to be responsible for creating/reusing a WebView when Automation asks for a page. The client might need to create a new application window, or even doing nothing if it's not possible.

haruhisa-shin replied:
I moved the WebView creation to the AutomationSessionClient.

Also, the following changes no longer allow the `NavigationActionData` to be initialized uniformly, so individual member initialization has been added.
https://github.com/WebKit/WebKit/commit/a5cb991f8f7b3d8c8a0362e958c5ab57043183b8
https://github.com/WebKit/WebKit/commit/392e336708452baf15ee7827968d2860a43391a5

burg commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -50,58 +50,25 @@ AutomationSessionClient::AutomationSessionClient(const String& sessionIdentifier
-    uiClient.base.version = 0;
-    uiClient.base.clientInfo = this;
-    uiClient.close = close;
-    WKPageSetPageUIClient(toAPI(newPage), &uiClient.base);
```
In the Cocoa port, clients get a chance to install their own delegates/mess with page configurations on WebDriver-created WebViews. It seems like this is hardcoded here, which could be problematic if you expect WebDriver WebViews to call normal PageUIClient and NavigationClient methods.

haruhisa-shin replied:
As you said. The hardcode here is not appropriate and has been removed.

fujii reviewed (approved):
LGTM

haruhisa-shin committed:
[WebDriver][win] make the `New Window` command duplicates an application window, instead of creating new one.
https://bugs.webkit.org/show_bug.cgi?id=289887

Reviewed by Fujii Hironori.

Before this change, the WebDriver creates new WebPageProxy and WebView for the `New Window` command.
This method is simple but not appropriate for testing the application, as the newly created window
differs from the application's.
The purpose of this change is to solve the above problem.

The main change in this commit is that the AutomationSessionClient will no longer create a WebView
when it receives `New Window` command, but will instead call the `WKPageUIClient::createNewPage`.
The creation of the window is delegated to the application, which will allow for more correct testing.

Combined changes:
* Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
(WebKit::AutomationClient::closeAutomationSession):
* Source/WebKit/UIProcess/win/AutomationSessionClientWin.cpp:
(WebKit::AutomationSessionClient::requestNewPageWithOptions):
(WebKit::getRootWindowHandle):
(WebKit::AutomationSessionClient::requestMaximizeWindowOfPage):
(WebKit::AutomationSessionClient::requestHideWindowOfPage):
(WebKit::AutomationSessionClient::requestRestoreWindowOfPage):
(WebKit::AutomationSessionClient::close): Deleted.
(WebKit::AutomationSessionClient::didReceiveAuthenticationChallenge): Deleted.
(WebKit::AutomationSessionClient::retainWebView): Deleted.
(WebKit::AutomationSessionClient::releaseWebView): Deleted.
* Source/WebKit/UIProcess/win/AutomationSessionClientWin.h:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
(WebKitBrowserWindow::WebKitBrowserWindow):
(WebKitBrowserWindow::didReceiveAuthenticationChallenge):
(WebKitBrowserWindow::close):
* Tools/MiniBrowser/win/WebKitBrowserWindow.h:

Canonical link: https://commits.webkit.org/293505@main
